File: web/src/lib/chat/slash-picker.ts
/**
 * Composer `/` picker model (slash-discovery). Catalog, phase, prefix
 * filter, insert strings, Dex-row merge, and bind-still-valid. Pure — no
 * I/O, no POST.
 */

import {
  parseSlashCommand,
  slashArg,
  type DexBind,
} from "./slash-commands";

export const SLASH_COMMANDS = [
  { token: "/new", hint: "New empty chat", trailingSpace: false, arg: "none" },
  {
    token: "/team",
    hint: "Open Teams",
    hintGuest: "Open Teams · sign in to save",
    trailingSpace: true,
    arg: "team",
  },
  { token: "/dex", hint: "Open Dex", trailingSpace: true, arg: "dex" },
  { token: "/usage", hint: "Open live usage", trailingSpace: true, arg: "usage" },
  { token: "/calc", hint: "Open calculator", trailingSpace: true, arg: "none" },
  { token: "/help", hint: "Show these commands", trailingSpace: false, arg: "none" },
] as const;

export type CommandRow = (typeof SLASH_COMMANDS)[number];

export type SlashPickerPhase =
  | { phase: "hidden" }
  | { phase: "commands"; prefix: string; rows: CommandRow[] }
  | { phase: "args"; command: "dex" | "team" | "usage"; query: string }
  | { phase: "rest"; command: "calc" | "new" | "help" };

export type DexNameRow = {
  kind: "pokemon" | "move" | "ability" | "item";
  slug: string;
  displayName: string;
  spriteUrl?: string;
};

export const PICKER_CAPTION = "Insert, then send";
export const EMPTY_DEX = "No Dex matches";
export const EMPTY_USAGE = "No usage matches";
export const EMPTY_TEAMS = "No saved teams match";
export const EMPTY_TEAMS_GUEST = "Sign in to save teams";

const KIND_ORDER: DexNameRow["kind"][] = ["pokemon", "move", "ability", "item"];

export function slashPickerPhase(text: string): SlashPickerPhase {
  const trimmed = text.trimStart();
  if (!trimmed.startsWith("/")) return { phase: "hidden" };

  const match = /^\S+/.exec(trimmed);
  const token = match ? match[0] : "";
  const hasSpaceAfter = /^\s/.test(trimmed.slice(token.length));

  if (!hasSpaceAfter) {
    const rows = filterCommands(token);
    if (rows.length === 0) return { phase: "hidden" };
    return { phase: "commands", prefix: token, rows };
  }

  const command = SLASH_COMMANDS.find(
    (row) => row.token === token.toLowerCase(),
  );
  if (!command) return { phase: "hidden" };

  if (command.arg === "dex" || command.arg === "team" || command.arg === "usage") {
    return { phase: "args", command: command.arg, query: slashArg(text) };
  }

  const rest = command.token.slice(1);
  if (rest === "calc" || rest === "new" || rest === "help") {
    return { phase: "rest", command: rest };
  }
  return { phase: "hidden" };
}

export function filterCommands(prefix: string): CommandRow[] {
  const needle = prefix.toLowerCase();
  return SLASH_COMMANDS.filter((row) =>
    row.token.toLowerCase().startsWith(needle),
  );
}

export function insertCommand(token: string): string {
  const row = SLASH_COMMANDS.find(
    (command) => command.token.toLowerCase() === token.toLowerCase(),
  );
  if (!row) return token;
  return row.trailingSpace ? `${row.token} ` : row.token;
}

export function insertName(commandToken: string, displayName: string): string {
  return `${commandToken} ${displayName}`;
}

export function mergeDexNameRows(
  byKind: { kind: DexNameRow["kind"]; matches: DexNameRow[] }[],
  limit = 8,
): DexNameRow[] {
  const buckets = new Map<DexNameRow["kind"], DexNameRow[]>();
  for (const kind of KIND_ORDER) buckets.set(kind, []);
  for (const group of byKind) {
    buckets.get(group.kind)?.push(...group.matches);
  }

  const seen = new Set<string>();
  const merged: DexNameRow[] = [];
  for (const kind of KIND_ORDER) {
    for (const row of buckets.get(kind) ?? []) {
      const key = `${row.kind}:${row.slug}`;
      if (seen.has(key)) continue;
      seen.add(key);
      merged.push(row);
      if (merged.length >= limit) return merged;
    }
  }
  return merged;
}

export function bindStillValid(bind: DexBind, composerText: string): boolean {
  const parsed = parseSlashCommand(composerText, { hasUsagePage: true });
  if (parsed.type !== "navigate" || parsed.target !== "dex") return false;
  return slashArg(composerText).toLowerCase() === bind.displayName.toLowerCase();
}

File: android/app/src/main/kotlin/ai/gowtam/oak/features/chat/SlashPicker.kt
package ai.gowtam.oak.features.chat

/**
 * Composer `/` picker model (slash-discovery). Catalog, phase, prefix
 * filter, insert strings, Dex-row merge, and bind-still-valid. Pure — no
 * I/O, no POST. Clone of `web/src/lib/chat/slash-picker.ts`.
 */

data class SlashCommandRow(
    val token: String,
    val hint: String,
    val hintGuest: String? = null,
    val trailingSpace: Boolean,
    val arg: String,
)

val SLASH_COMMANDS: List<SlashCommandRow> = listOf(
    SlashCommandRow(
        token = "/new",
        hint = "New empty chat",
        trailingSpace = false,
        arg = "none",
    ),
    SlashCommandRow(
        token = "/team",
        hint = "Open Teams",
        hintGuest = "Open Teams · sign in to save",
        trailingSpace = true,
        arg = "team",
    ),
    SlashCommandRow(
        token = "/dex",
        hint = "Open Dex",
        trailingSpace = true,
        arg = "dex",
    ),
    SlashCommandRow(
        token = "/usage",
        hint = "Open live usage",
        trailingSpace = true,
        arg = "usage",
    ),
    SlashCommandRow(
        token = "/calc",
        hint = "Open calculator",
        trailingSpace = true,
        arg = "calc",
    ),
    SlashCommandRow(
        token = "/help",
        hint = "Show these commands",
        trailingSpace = false,
        arg = "none",
    ),
)

sealed interface SlashPickerPhase {
    data object Hidden : SlashPickerPhase
    data class Commands(val prefix: String, val rows: List<SlashCommandRow>) : SlashPickerPhase
    data class Args(val command: String, val query: String) : SlashPickerPhase
    data class Rest(val command: String) : SlashPickerPhase
}

data class DexNameRow(
    val kind: DexNameKind,
    val slug: String,
    val displayName: String,
    val spriteUrl: String? = null,
)

data class DexKindMatches(
    val kind: DexNameKind,
    val matches: List<DexNameRow>,
)

const val PICKER_CAPTION = "Insert, then send"
const val EMPTY_DEX = "No Dex matches"
const val EMPTY_USAGE = "No usage matches"
const val EMPTY_TEAMS = "No saved teams match"
const val EMPTY_TEAMS_GUEST = "Sign in to save teams"

private val KIND_ORDER = listOf(
    DexNameKind.Pokemon,
    DexNameKind.Move,
    DexNameKind.Ability,
    DexNameKind.Item,
)

fun slashPickerPhase(text: String): SlashPickerPhase {
    val trimmed = text.trimStart()
    if (!trimmed.startsWith("/")) return SlashPickerPhase.Hidden

    val token = Regex("^\\S+").find(trimmed)?.value.orEmpty()
    val hasSpaceAfter = trimmed.substring(token.length).firstOrNull()?.isWhitespace() == true

    if (!hasSpaceAfter) {
        val rows = filterCommands(token)
        if (rows.isEmpty()) return SlashPickerPhase.Hidden
        return SlashPickerPhase.Commands(prefix = token, rows = rows)
    }

    val command = SLASH_COMMANDS.find { it.token == token.lowercase() }
        ?: return SlashPickerPhase.Hidden

    if (command.arg == "dex" || command.arg == "team" || command.arg == "usage" || command.arg == "calc") {
        return SlashPickerPhase.Args(command = command.arg, query = slashArgs(text))
    }

    val rest = command.token.removePrefix("/")
    if (rest == "new" || rest == "help") {
        return SlashPickerPhase.Rest(command = rest)
    }
    return SlashPickerPhase.Hidden
}

fun filterCommands(prefix: String): List<SlashCommandRow> {
    val needle = prefix.lowercase()
    return SLASH_COMMANDS.filter { it.token.lowercase().startsWith(needle) }
}

fun insertCommand(token: String): String {
    val row = SLASH_COMMANDS.find { it.token.equals(token, ignoreCase = true) } ?: return token
    return if (row.trailingSpace) "${row.token} " else row.token
}

fun insertName(commandToken: String, displayName: String): String = "$commandToken $displayName"

fun mergeDexNameRows(
    byKind: List<DexKindMatches>,
    limit: Int = 8,
): List<DexNameRow> {
    val buckets = KIND_ORDER.associateWith { mutableListOf<DexNameRow>() }.toMutableMap()
    for (group in byKind) {
        buckets[group.kind]?.addAll(group.matches)
    }

    val seen = mutableSetOf<String>()
    val merged = mutableListOf<DexNameRow>()
    for (kind in KIND_ORDER) {
        for (row in buckets[kind].orEmpty()) {
            val key = "${row.kind}:${row.slug}"
            if (key in seen) continue
            seen += key
            merged += row
            if (merged.size >= limit) return merged
        }
    }
    return merged
}

fun bindStillValid(bind: DexBind, composerText: String): Boolean {
    val parsed = parseSlashCommand(composerText, hasUsagePage = true)
    if (parsed !is SlashCommand.Navigate || parsed.target != SlashCommand.Target.Dex) return false
    return slashArgs(composerText).equals(bind.displayName, ignoreCase = true)
}

fun DexNameKind.label(): String = when (this) {
    DexNameKind.Pokemon -> "Pokémon"
    DexNameKind.Move -> "Move"
    DexNameKind.Ability -> "Ability"
    DexNameKind.Item -> "Item"
}
